src/transformers/models/llama/modeling_llama.py

```diff
@@ -263,34 +293,27 @@ def __init__(self, config: LlamaConfig, layer_idx: Optional[int] = None):
         self.k_proj = nn.Linear(self.hidden_size, self.num_key_value_heads * self.head_dim, bias=config.attention_bias)
         self.v_proj = nn.Linear(self.hidden_size, self.num_key_value_heads * self.head_dim, bias=config.attention_bias)
         self.o_proj = nn.Linear(self.hidden_size, self.hidden_size, bias=config.attention_bias)
-        self._init_rope()
-
-    def _init_rope(self):
-        if self.config.rope_scaling is None:
-            self.rotary_emb = LlamaRotaryEmbedding(
-                self.head_dim,
-                max_position_embeddings=self.max_position_embeddings,
-                base=self.rope_theta,
-            )
-        else:
-            scaling_type = self.config.rope_scaling["type"]
-            scaling_factor = self.config.rope_scaling["factor"]
-            if scaling_type == "linear":
-                self.rotary_emb = LlamaLinearScalingRotaryEmbedding(
-                    self.head_dim,
-                    max_position_embeddings=self.max_position_embeddings,
-                    scaling_factor=scaling_factor,
-                    base=self.rope_theta,
-                )
-            elif scaling_type == "dynamic":
-                self.rotary_emb = LlamaDynamicNTKScalingRotaryEmbedding(
-                    self.head_dim,
-                    max_position_embeddings=self.max_position_embeddings,
-                    scaling_factor=scaling_factor,
-                    base=self.rope_theta,
-                )
-            else:
-                raise ValueError(f"Unknown RoPE scaling type {scaling_type}")
+
+        rope_config = {
+            "dim": self.head_dim,
+            "max_position_embeddings": self.max_position_embeddings,
+            "base": self.rope_theta,
+            "rope_type": "default",
+        }
+
+        if self.config.rope_scaling is not None:
+            rope_config["rope_type"] = self.config.rope_scaling["type"]  # overwrites '"rope_type": "default"'
+            rope_config["scaling_factor"] = self.config.rope_scaling["factor"]
+            if "yarn" in rope_config["rope_type"]:
+                for yarn_parameter in [
+                    "original_max_position_embeddings",
+                    "attention_factor",
+                    "beta_fast",
+                    "beta_slow",
+                ]:
+                    rope_config[yarn_parameter] = self.config.rope_scaling.get(yarn_parameter)
+
+        self.rotary_emb = LlamaRotaryEmbedding(**rope_config)
```

Comment:
we still have a ROBE layer at each attention layer, which we should get rid off IMO